lib/wibox/layout/ratio.lua

```diff
@@ -0,0 +1,339 @@
+local function gen_sum(self, i_s, i_e)
+    local sum, new_w = 0,0
+
+    for i = i_s or 1, i_e or #self.widgets do
+        if self._ratios[i] then
+            sum = sum + self._ratios[i]
+        else
+            new_w = new_w + 1
+        end
+    end
+
+    return sum, new_w
+end
+
+--- Make sure "diff" the sum is 1
+local function normalize(self)
+    local count = #self.widgets
+    if count == 0 then return end
+
+    -- Instead of adding "if" everywhere, just handle this common case
+    if count == 1 then
+        self._ratios = { 1 }
+        return
+    end
+
+    local sum, new_w = gen_sum(self)
+    local old_count  = #self.widgets - new_w
+
+    local to_add = (sum == 0) and 1 or (sum / old_count)
+
+    -- Make sure all widgets have a ratio
+    for i=1, #self.widgets do
+        if not self._ratios[i] then
+            self._ratios[i] = to_add
+        end
+    end
+
+    sum = sum + to_add*new_w
+
+    local delta, new_sum =  (1 - sum) / count,0
+
+    -- Increase or decrease each ratio so it the sum become 1
+    for i=1, #self.widgets do
+        self._ratios[i] = self._ratios[i] + delta
+        new_sum = new_sum + self._ratios[i]
+    end
+
+    -- Floating points summation isn't an exact science, worst on 32bit systems
+    assert(new_sum > 0.99 and new_sum < 1.01)
+end
+
+--- Layout a ratio layout. Each widget gets a share of the size proportional
+-- to its ratio
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function ratio:layout(context, width, height)
+    local result = {}
+    local pos,spacing = 0, self._spacing
+    local num = #self.widgets
+    local total_spacing = (spacing*(num-1))
+
+    normalize(self)
+
+    for k, v in ipairs(self.widgets) do
+        local space = nil
+        local x, y, w, h
+
+        if self.dir == "y" then
+            space = height * self._ratios[k]
+            x, y = 0, util.round(pos)
+            w, h = width, floor(space)
+        else
+            space = width * self._ratios[k]
+            x, y = util.round(pos), 0
+            w, h = floor(space), height
+        end
+
+        table.insert(result, base.place_widget_at(v, x, y, w, h))
+
+        pos = pos + space + spacing
+
+        if (self.dir == "y" and pos-spacing >= height) or
+            (self.dir ~= "y" and pos-spacing >= width) then
+            break
+        end
+    end
+
+    return result
+end
+
+--- Increase the ratio of "widget"
+-- If the increment produce an invalid ratio (not between 0 and 1), the method
+-- do nothing.
+-- @tparam number index The widget index to change
+-- @tparam number increment An floating point value between -1 and 1 where the
+--   end result is within 0 and 1
+function ratio:inc_ratio(index, increment)
+    if #self.widgets ==  1 or (not index) or increment < -1 or increment > 1 then return end
+
+    local widget = self.widgets[index]
+
+    assert(self._ratios[index])
```

Comment:
Why does this check if `index` is a valid index with an assert? In other code you used `return false` for this case.